lib/charms/mongodb/v1/mongos.py

```diff
@@ -171,30 +175,44 @@ def remove_shard(self, shard_name: str) -> None:
 
         Raises:
             ConfigurationError, OperationFailure, NotReadyError,
-            RemovePrimaryShardError
+            NotEnoughSpaceError
         """
-        sc_status = self.client.admin.command("listShards")
-
         # It is necessary to call removeShard multiple times on a shard to guarantee removal.
         # Allow re-removal of shards that are currently draining.
+        sc_status = self.client.admin.command("listShards")
         if self._is_any_draining(sc_status, ignore_shard=shard_name):
             cannot_remove_shard = (
                 f"cannot remove shard {shard_name} from cluster, another shard is draining"
             )
             logger.error(cannot_remove_shard)
             raise NotReadyError(cannot_remove_shard)
 
-        databases_using_shard_as_primary = self.get_databases_for_shard(shard_name)
-        if databases_using_shard_as_primary:
-            cannot_remove_primary_shard = f"These databases: {', '.join(databases_using_shard_as_primary)}, use Shard {shard_name} is a primary shard, cannot remove shard."
-            logger.error(cannot_remove_primary_shard)
-            raise RemovePrimaryShardError(cannot_remove_primary_shard)
-
+        # remove and process removal status
         logger.info("Attempting to remove shard %s", shard_name)
         removal_info = self.client.admin.command("removeShard", shard_name)
+        self._log_removal_info(removal_info, shard_name)
+        remaining_chunks = self._retrieve_remaining_chunks(removal_info)
+        if remaining_chunks:
+            logger.info("Waiting for all chunks to be drained from %s.", shard_name)
+            raise NotDrainedError()
+
+        # MongoDB docs says to movePrimary only after all chunks have been drained from the shard.
+        logger.info("All chunks drained from shard: %s", shard_name)
+        databases_using_shard_as_primary = self.get_databases_for_shard(shard_name)
+        if not databases_using_shard_as_primary:
+            return
```

Comment:
Should we call `removeShard` again here?

Reply:
There are to cases here:
1. primaries to move
2. no primaries to move

in case 1, we should not, we should move primaries and **then** call removeShard. Which we do
in case 2, if the shard is not fully drained an error was raised on the previous lines and the code will be deferred and attempted again.

FYI draining shards can take several days depending how much data is present, so IMO deferring isn't detrimental here